crates/next-dev/src/lib.rs

```diff
@@ -441,3 +406,105 @@ pub async fn start_server(options: &DevServerOptions) -> Result<()> {
 
     Ok(())
 }
+
+#[allow(clippy::too_many_arguments)]
+#[turbo_tasks::function]
+pub async fn get_root_source(
+    project_path: FileSystemPathVc,
+    output_root: FileSystemPathVc,
+    dev_server_root: FileSystemPathVc,
+    env: ProcessEnvVc,
+    browserslist_query: StringVc,
+    eager_compile: BoolVc,
+    server_component_externals: StringsVc,
+    entry_requests: Vec<RequestVc>,
+    viz: ContentSourceVc,
+) -> Result<ContentSourceVc> {
+    let browserslist_query = &*browserslist_query.await?;
+    let eager_compile = *eager_compile.await?;
+
+    let app_source_vc = create_app_source(
+        project_path,
+        output_root.join("app"),
+        dev_server_root,
+        env,
+        browserslist_query,
+        server_component_externals,
+    );
+    let app_source = *app_source_vc.await?;
+
+    let web_source_vc = create_web_entry_source(
+        project_path,
+        entry_requests,
+        dev_server_root,
+        env,
+        eager_compile,
+        browserslist_query,
+    );
+    let web_source = *web_source_vc.await?;
+
+    let rendered_source_vc = create_server_rendered_source(
+        project_path,
+        output_root.join("pages"),
+        dev_server_root,
+        env,
+        browserslist_query,
+    );
+    let rendered_source = *rendered_source_vc.await?;
+
+    if web_source.is_none() && rendered_source.is_none() && app_source.is_none() {
+        NextAppIssue {
+            severity: IssueSeverity::Error.into(),
+            path: project_path,
+            message: StringVc::cell(
+                "Missing Next.js entry file(s): src/index.js, pages/*.js, and app/**/page.js are \
+                 missing."
+                    .into(),
+            ),
+        }
+        .cell()
+        .as_issue()
+        .emit();
+    }
+
+    let web_source = match web_source {
+        Some(source) => source,
+        None => NoContentSourceVc::new().into(),
+    };
```

Comment:
Nit: 
```suggestion
    let web_source = web_source.map(|| NoContentSourceVc::new().into());
```

Reply:
I had the same thought, but: doesn't `map` map the value in the `Some()` case? We kind of need something like an equivalent of `Result`'s `map_err` — "map_none" would be nice: https://internals.rust-lang.org/t/should-we-add-map-none-method-for-option-type/15859

Reply:
Ha, it's `unwrap_or_else` that I meant to type.

Reply:
Ha, it's [unwrap_or_else](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else) that I meant to type.